filament/backend/test/test_FeedbackLoops.cpp

```diff
@@ -66,132 +120,128 @@ TEST_F(BackendTest, FeedbackLoops) {
-                {},                                        // depth
-                {});                                       // stencil
-
-        // Fill the texture with interesting colors.
-        const size_t size = 512 * 512 * 4;
+            SamplerType::SAMPLER_2D, kNumLevels, kTexFormat, 1, kTexWidth, kTexHeight, 1, usage);
+
+        // Create a RenderTarget for each miplevel.
+        Handle<HwRenderTarget> renderTargets[kNumLevels];
+        for (uint8_t level = 0; level < kNumLevels; level++) {
+            printf("Level %d: %d x %d\n", level, kTexWidth >> level, kTexHeight >> level);
+            renderTargets[level] = getDriverApi().createRenderTarget( TargetBufferFlags::COLOR,
+                    kTexWidth >> level, kTexHeight >> level, 1, { texture, level, 0 }, {}, {});
+        }
+
+        // Fill the base level of the texture with interesting colors.
+        const size_t size = kTexHeight * kTexWidth * 4;
         uint8_t* buffer = (uint8_t*) malloc(size);
-        for (int r = 0, i = 0; r < 512; r++) {
-            for (int c = 0; c < 512; c++, i += 4) {
+        for (int r = 0, i = 0; r < kTexHeight; r++) {
+            for (int c = 0; c < kTexWidth; c++, i += 4) {
                 buffer[i + 0] = 0x10;
-                buffer[i + 1] = 0xff * r / 511;
-                buffer[i + 2] = 0xff * c / 511;
+                buffer[i + 1] = 0x1f * r / (kTexHeight - 1);
+                buffer[i + 2] = 0x1f * c / (kTexWidth - 1);
                 buffer[i + 3] = 0xf0;
             }
          }
         auto cb = [](void* buffer, size_t size, void* user) { free(buffer); };
         PixelBufferDescriptor pb(buffer, size, PixelDataFormat::RGBA, PixelDataType::UBYTE, cb);
+        getDriverApi().update2DImage(texture, 0, 0, 0, kTexWidth, kTexHeight, std::move(pb));
+
+        for (int frame = 0; frame < kNumFrames; frame++) {
+
+            // Prep for rendering.
+            RenderPassParams params = {};
+            params.flags.clear = TargetBufferFlags::NONE;
+            params.flags.discardEnd = TargetBufferFlags::NONE;
+            PipelineState state;
+            state.rasterState.colorWrite = true;
+            state.rasterState.depthWrite = false;
+            state.rasterState.depthFunc = RasterState::DepthFunc::A;
+            state.program = program;
+            backend::SamplerGroup samplers(1);
+            backend::SamplerParams sparams = {};
+            sparams.filterMag = SamplerMagFilter::LINEAR;
+            sparams.filterMin = SamplerMinFilter::LINEAR_MIPMAP_NEAREST;
+            samplers.setSampler(0, texture, sparams);
+            auto sgroup = getDriverApi().createSamplerGroup(samplers.getSize());
+            getDriverApi().updateSamplerGroup(sgroup, std::move(samplers.toCommandStream()));
+            auto ubuffer = getDriverApi().createUniformBuffer(sizeof(MaterialParams),
+                    backend::BufferUsage::STATIC);
+            getDriverApi().makeCurrent(swapChain, swapChain);
```

Comment:
Maybe just create a local reference to the driver API object so you don't call the method all over the place? It would be just a bit cleaner